Initial registry sync implementation

Patbox opened the pull request:
Initial implementation of registry sync for qsl. 

While it is functional, I'm still not sure about api structure.
Other thing to consider, is proper blockstate sync, which isn't implemented yet.

Additionally I though about usage of small hack for having sync before play stage on server, but on it on client for nicer experience while keeping support for proxy software like velocity.

Patbox committed:
Initial registry sync implementation

LambdAurora commented:
You might want to execute `./gradlew :core:registry:applyLicenses`.

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/client/MinecraftClientMixin.java:
```diff
@@ -0,0 +1,24 @@
+@Mixin(MinecraftClient.class)
+public class MinecraftClientMixin {
+	@Inject(method = "disconnect(Lnet/minecraft/client/gui/screen/Screen;)V", at = @At("TAIL"))
+	private void polymer_onDisconnect(CallbackInfo ci) {
```
This looks suspicious :p

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/client/MinecraftClientMixin.java:
```diff
@@ -0,0 +1,24 @@
+	private void polymer_onDisconnect(CallbackInfo ci) {
+		for (var reg : Registry.REGISTRIES) {
+			if (reg instanceof SynchronizedRegistry registry && registry.quilt$requiresSyncing()) {
+				System.out.println("Restored snapshot of " + ((Registry<Registry<?>>) Registry.REGISTRIES).getId(reg));
```
I don't think this is wanted.

Patbox committed:
Run applyLicenses

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/client/ConnectScreenMixin.java:
```diff
@@ -0,0 +1,24 @@
+	private void quilt$snapshotRegistry(MinecraftClient client, ServerAddress address, CallbackInfo ci) {
+		for (var reg : Registry.REGISTRIES) {
+			if (reg instanceof SynchronizedRegistry registry && registry.quilt$requiresSyncing()) {
+				System.out.println("Created snapshot of " + ((Registry<Registry<?>>) Registry.REGISTRIES).getId(reg));
```
Mmh, leftover debug print.

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/SimpleRegistryMixin.java:
```diff
@@ -51,8 +90,121 @@ protected SimpleRegistryMixin(RegistryKey<? extends Registry<V>> key, Lifecycle
 	)
 	private void quilt$invokeEntryAddEvent(int rawId, RegistryKey<V> key, V entry, Lifecycle lifecycle, boolean checkDuplicateKeys,
-	                                       CallbackInfoReturnable<Holder<V>> cir) {
+										   CallbackInfoReturnable<Holder<V>> cir) {
```
Extremely nit-picky but I would keep the spaces from smart-tabs as it keeps alignment.

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/BlocksMixin.java:
```diff
@@ -0,0 +1,35 @@
+	@Inject(method = "<clinit>", at = @At("RETURN"))
+	private static void onInit(CallbackInfo ci) {
+		RegistryMonitor.create(Registry.BLOCK)
+				.forUpcoming(context -> context.value().getStateManager().getStates().forEach(Block.STATE_IDS::add));
```
If you use forUpcoming you can use the event directly?

Not sure which way should be the default.

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,46 @@
+
+	/**
+	 * Sets registry next data packets belong to
+	 * {
```
Code stuff like this probably should be wrapped around `<pre><code>` blocks.

Patbox committed:
Cleanup

OroArmor commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/RegistryLibInitializer.java:
```diff
@@ -0,0 +1,31 @@
+import org.quiltmc.qsl.registry.impl.sync.ClientRegistrySync;
+
+@Environment(EnvType.CLIENT)
+public class RegistryLibInitializer implements ClientModInitializer {
```
```suggestion
public class RegistrySyncClientInitializer implements ClientModInitializer {
```
Dont forget the file name too

OroArmor commented on library/core/registry/src/testmod/java/org/quiltmc/qsl/registry/test/RegistryLibSyncTest.java:
```diff
@@ -0,0 +1,100 @@
+import java.nio.file.StandardOpenOption;
+
+
+public class RegistryLibSyncTest implements ModInitializer {
```
I would make a comment or two about what this test does. It seems to make sure there is a client server mismatch, but I'm not 100% sure

LambdAurora replied:
I also would add `@ApiStatus.Internal` to initializers.

Patbox replied:
It forces rawId mismatch to make sure it works correctly (with identifiers being the same, since it's base for sync)

Patbox replied:
> @ApiStatus.Internal

Not sure if it's needed there, since it's part of test mod

Patbox committed:
Fix some issues, don't sync vanilla/unmodified registries

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

# Conflicts:
#	library/core/registry/src/main/resources/fabric.mod.json

Patbox committed:
Improve api, make client respond and send information about protocols version

TheGlitch76 commented:
Hi, I will probably not get to this until after the Beta launch, but I do intend to review this ASAP.

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

Patbox committed:
Update initializers

TropheusJ commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	/**
+	 * Blocks entry from being synced
```
took me a good 30 seconds to realize this means `Blocks` as in `prevents` and not a Minecraft block - maybe change to `Prevents`?

TropheusJ commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	/**
+	 * Blocks entry from being synced
```
here too

TropheusJ commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java:
```diff
@@ -0,0 +1,186 @@
+		} else if (RegistryFlag.isOptional(flags)) {
+			optionalRegistry = true;
+		} else {
+			LOGGER.warn("Trying to sync registry " + identifier + " which doesn't " + (registry == null ? "support it!" : "exist!"));
```
swapped logic on this ternary?

Patbox committed:
Fix reversed ternary

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,101 @@
+public final class RegistrySynchronization {
+
+	/**
+	 * Marks registry status as dirty, rebuilding its sync data
```
The javadoc says that this method rebuilds sync data, but the implementation does no such thing!
```suggestion
	 * Marks registry status as dirty, requiring its synchronization data to be rebuilt.
```

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Identifiers of packets send by server
```
```suggestion
 * Identifiers of packets sent by server
```

TheGlitch76 commented on library/core/registry/src/testmod/java/org/quiltmc/qsl/registry/test/RegistryLibSyncTest.java:
```diff
@@ -0,0 +1,112 @@
+ * Items/Blocks are registered in different order on client/server to make sure sync works correctly
+ * Server also gets its own entry that shouldn't block client from joining
+ */
+public class RegistryLibSyncTest implements ModInitializer {
```
There is a lot of stuff in this API that needs covered by tests:
- Registry sync for registries added by mods
- Optional entries and registries
- Skipped entries and registries

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ClientPackets.java:
```diff
@@ -0,0 +1,53 @@
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Identifiers of packets send by server
```
```suggestion
 * Identifiers of packets sent by server
```

BasiqueEvangelist commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,101 @@
+	 * Makes registry synchronize with client
+	 */
+	public static void synchronize(SimpleRegistry<?> registry) {
+		SynchronizedRegistry.as(registry).quilt$requiresSyncing();
```
Doesn't `quilt$requiresSyncing` just return a boolean?
```suggestion
		SynchronizedRegistry.as(registry).quilt$markForSync();
```

Patbox replied:
Ah right, somehow missed that

Patbox committed:
Update library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java

Co-authored-by: BasiqueEvangelist <basiqueevangelist@yandex.ru>

BasiqueEvangelist commented:
Could you add a method to the API which resyncs registries with a player? I'm using an FAPI impl method in DynReg right now, but that won't work with QSL...

Also, are you going to put saving registries to disk in a different PR?

Patbox commented:
> Could you add a method to the API which resyncs registries with a player? I'm using an FAPI impl method in DynReg right now, but that won't work with QSL...
> 
> Also, are you going to put saving registries to disk in a different PR?

Sounds like something I could add. My long term I plan on moving registry sync to it's own "virtual" networking stage (kinda like what polymer does currently) to allow nicer interaction with vanilla clients (and possibly fabric ones). However I need to make api for that first

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

# Conflicts:
#	library/core/registry/src/testmod/resources/fabric.mod.json

Patbox commented:
For saving on disk, is it really needed? Outside of Status Effect (which I might pr alternative/proper fix later), I don't think there is real need for it

Patbox committed:
Merge remote-tracking branch 'patbox/1.18' into 1.18

BasiqueEvangelist commented:
> For saving on disk, is it really needed? Outside of Status Effect (which I might pr alternative/proper fix later), I don't think there is real need for it

Ah right, they removed the need for biomes to be persisted, I forgot about that.

> Sounds like something I could add. My long term I plan on moving registry sync to it's own "virtual" networking stage (kinda like what polymer does currently) to allow nicer interaction with vanilla clients (and possibly fabric ones). However I need to make api for that first

Nice!

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * Makes registry synchronize with client
+	 */
+	public static void synchronize(SimpleRegistry<?> registry) {
```
i think markForSync makes more sense as a name, and that seems to be what you use in the rest of the codebase.

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/RegistryFlag.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Registry flags used to determine how registry (entries) should be handled by sync
+ */
+@ApiStatus.Experimental
```
No need to mark something that isn't API as experimental API

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+ * Identifiers of packets send by server
+ */
+@ApiStatus.Internal
+public final class ServerPackets {
```
IMO these channels could all use more descriptive names

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	 * }
+	 * </code></pre>
+	 */
+	public static final Identifier HELLO = id("registry/hello");
```
registry_sync/handshake
since this is where the client + server negotiate the version of registry sync to use

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	/**
+	 * Ends registry sync. No data
+	 */
+	public static final Identifier GOODBYE = id("registry/goodbye");
```
registry_sync/finish?

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	 * }
+	 * </code></pre>
+	 */
+	public static final Identifier START = id("registry/start");
```
registry_sync/begin_registry?

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	 * <pre><code>
+	 * {
+	 *   Registry identifier: Identifier
+	 *   Count: VarInt
```
count of what?

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	 * }
+	 * </code></pre>
+	 */
+	public static final Identifier DATA = id("registry/data");
```
registry_sync/registry_data

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	/**
+	 * Applies changes, doesn't have any data
+	 */
+	public static final Identifier APPLY = id("registry/apply");
```
registry_sync/apply

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+	public static final Identifier DATA = id("registry/data");
+
+	/**
+	 * Applies changes, doesn't have any data
```
Does this mark the end of one START cycle, or is it used to apply all registry changes at once? Could use clarification in javadoc.

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ClientPackets.java:
```diff
@@ -0,0 +1,53 @@
+	 * Sent after registry sync failure before client disconnect
+	 * <pre><code>
+	 * {
+	 *     Registry: Identifier
```
any reason to include this if its seemingly always ignored?

Patbox replied:
In future (aka when I get around virtual networking stages api), kicking will be moved to server. Through I guess for now it can be removed and readded in later sync version

TheGlitch76 replied:
It's fine to future-proof and keep it in.

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerRegistrySync.java:
```diff
@@ -0,0 +1,162 @@
+	// Ideally we should introduce new stage/api for it by creating custom ServerPlayPacketListener
+	// that just handles all of custom logic/emulation of login stage, which is fairly possible
+	// to do while keeping compatibility with vanilla clients (as long as other mods on server allow it).
+	// It's thanks to ping packets send by server and tpc/Minecraft preserving packet order!
```
what's tpc?

Patbox replied:
TCP, just got letters switched

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerRegistrySync.java:
```diff
@@ -0,0 +1,162 @@
+	// Ideally we should introduce new stage/api for it by creating custom ServerPlayPacketListener
+	// that just handles all of custom logic/emulation of login stage, which is fairly possible
+	// to do while keeping compatibility with vanilla clients (as long as other mods on server allow it).
+	// It's thanks to ping packets send by server and tpc/Minecraft preserving packet order!
```
```suggestion
	// It's thanks to ping packets sent by server and tpc/Minecraft preserving packet order!
```

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ClientPackets.java:
```diff
@@ -0,0 +1,53 @@
+@ApiStatus.Internal
+public final class ClientPackets {
+	/**
+	 * Response for {@link ServerPackets#HELLO}
```
```suggestion
	 * Response for {@link ServerPackets#HELLO}. Selects the registry sync version to be used from the server's supported options.
```

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerPackets.java:
```diff
@@ -0,0 +1,87 @@
+ * Identifiers of packets send by server
+ */
+@ApiStatus.Internal
+public final class ServerPackets {
```
Are we sure we want to keep the registry sync packet structure an impl detail, instead of considering it part of the public API?

Patbox replied:
Well structure could be part of public, but ideally mods shouldn't need it anyway. I guess documenting it somewhere is something we could look

TheGlitch76 commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/RegistryMixin.java:
```diff
@@ -62,4 +63,25 @@ private static void onFreezeBuiltins(CallbackInfo ci) {
+
+	@Inject(method = "<clinit>", at = @At("TAIL"))
+	private static void quilt$markRegistriesForSync(CallbackInfo ci) {
+		SynchronizedRegistry.markForSync(
```
Isn't it kind of unexpected behavior for new registries to not be synced by default?
Maybe a better solution is to sync all by default, and use RegistrySynchronization as an opt-out instead?

Patbox replied:
Vanilla ones are also opt in, so it made sense to me to apply that globally, but I can change that

Patbox committed:
Apply suggestions from code review

Co-authored-by: Glitch <glitch.g3431@gmail.com>

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

Patbox committed:
Apply rest of suggestions

Patbox committed:
Merge remote-tracking branch 'patbox/1.18' into 1.18

# Conflicts:
#	library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ClientPackets.java
#	library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerRegistrySync.java

TheGlitch76 reviewed (approved):
I feel pretty comfortable with this code now. Nice work!

TheGlitch76 committed:
Merge branch '1.18' into 1.18

TheGlitch76 committed:
Merge branch '1.18' into 1.18

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

Patbox committed:
Make sync run early, add detection for vanilla clients, add support for legacy/Fabric API sync protocol

Patbox commented:
Ok updated this pr with:
- Moved registry sync to run earlier to prevent clients from connecting when they are missing required entries/use vanilla
- Added detection of vanilla client with customisable kick message
- Added support for Fabric Registry Sync directly into this module

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+import org.quiltmc.qsl.registry.impl.sync.SynchronizedRegistry;
+
+/**
+ * Methods for manipulation of registry sync
```
```suggestion
 * Methods for manipulation of registry synchronization.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Makes registry synchronize with client
```
```suggestion
	 * Makes registry synchronize with clients.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Makes registry optional for sync
```
```suggestion
	 * Marks a registry as optional for synchronization.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Checks if registry is optional
```
```suggestion
	 * {@return {@code true} if the given registry is marked as optional, or {@code false} otherwise}
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Makes registry entry optional for sync
```
```suggestion
	 * Marks the given registry entry as optional of synchronization.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Makes registry entry optional for sync
```
```suggestion
	 * Marks the specified registry entry as optional of synchronization.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Checks if registry is optional
```
```suggestion
	 * {@return {@code true} if the registry entry is marked as optional, or {@code false} otherwise}
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Prevents entry from being synced
```
```suggestion
	 * Prevents the given registry entry from being synced.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Prevents entry from being synced
```
```suggestion
	 * Prevents the specified registry entry from being synced.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/sync/RegistrySynchronization.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	/**
+	 * Checks if registry is skipped from sync
```
```suggestion
	 * {@return {@code true} if the given registry entry is skipped in synchronization, or {@code false} otherwise}
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/RegistryConfig.java:
```diff
@@ -0,0 +1,39 @@
+import org.quiltmc.config.api.values.TrackedValue;
+import org.quiltmc.loader.api.config.QuiltConfig;
+
+public class RegistryConfig {
```
```suggestion
@ApiStatus.Internal
public class RegistryConfig {
```

I also would appreciate more blank lines in this class, everything is so close together right now.

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ClientPackets.java:
```diff
@@ -0,0 +1,53 @@
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Identifiers of packets sent by server
```
```suggestion
 * Identifiers of packets sent by server.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/RegistryFlag.java:
```diff
@@ -0,0 +1,50 @@
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Registry flags used to determine how registry (entries) should be handled by sync
```
```suggestion
 * Registry flags used to determine how registry (entries) should be handled by sync.
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerRegistrySyncNetworkHandler.java:
```diff
@@ -0,0 +1,349 @@
+ * protocol isn't too hard and gives equal results. And doing them on PLAY
+ * is needed for Fabric compatibility anyway
+ */
+public final class ServerRegistrySyncNetworkHandler implements ServerPlayPacketListener {
```
```suggestion
@ApiStatus.Internal
public final class ServerRegistrySyncNetworkHandler implements ServerPlayPacketListener {
```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java:
```diff
@@ -0,0 +1,260 @@
+
+	private static void handleStartPacket(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender sender) {
+		var identifier = buf.readIdentifier();
+		var count = buf.readVarInt();
```
Please don't use `var` for primitives.

Patbox committed:
Apply suggestions from code review

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Patbox committed:
Don't use var on primitives, add missing imports

Patbox committed:
Merge remote-tracking branch 'origin/1.18' into 1.18

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/RegistryConfig.java:
```diff
@@ -0,0 +1,41 @@
+	public static final String NO_REGISTRY_SYNC_MESSAGE = "no_registry_sync_message";
+	public static final String SUPPORT_FABRIC_API_SYNC = "support_fabric_api_sync";
+	private static Config config = null;
+	public static Config getConfig() {
```
The way that the config system is implemented is way too cursed; I'd recommend basing it on Ok Zoomer's [Config](https://github.com/EnnuiL/OkZoomer/blob/1.19/src/main/java/io/github/ennuil/ok_zoomer/config/OkZoomerConfig.java)/[ConfigManager](https://github.com/EnnuiL/OkZoomer/blob/1.19/src/main/java/io/github/ennuil/ok_zoomer/config/OkZoomerConfigManager.java) system, or even Haven's [basic Config example](https://github.com/Haven-King/developer-wiki/blob/main/wiki/configs/basic/markdown/basic.md), considering that the only thing being done with the config entries is reading it
For keeping snake_case on these examples (which I do recommend, please don't use camelCase on TOML), there's still the cursedness of using it on few fields, but we can fix it later once naming conventions are here

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/ServerRegistrySyncNetworkHandler.java:
```diff
@@ -0,0 +1,351 @@
+
+
+	@Override
+	public void onHandSwing(HandSwingC2SPacket packet) {
```
There are *way* too many blank lines there; Change all of those to `{ }`

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/IdListMixin.java:
```diff
@@ -0,0 +1,48 @@
+
+@Mixin(IdList.class)
+public class IdListMixin<T> implements SynchronizedIdList {
+
```
Make this go poof...?
```suggestion
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java:
```diff
@@ -0,0 +1,260 @@
+				registry.quilt$restoreIdSnapshot();
+			}
+		}
+		ClientRegistrySync.rebuildBlockStates();
```
I noticed that there are bits of static calls that has the class bit in (like this one) and many others without it; I'd recommend going through them all and add them back when it makes sense

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java:
```diff
@@ -0,0 +1,260 @@
+		} else {
+			LOGGER.warn("Trying to sync registry " + identifier + " which doesn't " + (registry != null ? "support it!" : "exist!"));
+			sendSyncFailedPacket(handler, identifier);
+			handler.getConnection().disconnect(getMessage("missing_registry","Client is missing required registry! Mismatched mods?"));
```
```suggestion
			handler.getConnection().disconnect(getMessage("missing_registry", "Client is missing required registry! Mismatched mods?"));
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java:
```diff
@@ -0,0 +1,260 @@
+		ClientPlayNetworking.registerGlobalReceiver(ServerPackets.REGISTRY_RESTORE, ClientRegistrySync::handleApplyPacket);
+		ClientPlayNetworking.registerGlobalReceiver(ServerPackets.END, ClientRegistrySync::handleGoodbyePacket);
+		ClientPlayNetworking.registerGlobalReceiver(ServerPackets.REGISTRY_RESTORE, ClientRegistrySync::handleRestorePacket);
+
```
Line go poof
```suggestion
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientFabricRegistrySync.java:
```diff
@@ -0,0 +1,185 @@
+				}
+			}
+		}
+
```
Yet another excess of lines
```suggestion
```

EnnuiL reviewed (changes requested):
Just address these, and all should be fine

Patbox committed:
Apply suggestions from code review

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

Patbox committed:
Apply rest of requested changes

EnnuiL reviewed (approved):
That looks much better now! I'd say only stylistic nitpicks are left, however, the logic itself seems fine; PR approved

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/RegistryConfig.java:
```diff
@@ -0,0 +1,36 @@
+
+@ApiStatus.Internal
+public class RegistryConfig extends WrappedConfig {
+	public static final RegistryConfig INSTANCE = QuiltConfig.create("quilt", "registry", RegistryConfig.class);
```
Let's define a standard here already: `qsl` as the containing folder, module name for the relevant config;
This is what Key Binds API was using, and it would be better in the long-run
```suggestion
	public static final RegistryConfig INSTANCE = QuiltConfig.create("qsl", "registry_sync", RegistryConfig.class);
```

LambdAurora replied:
I would keep the containing folder as quilt.

EnnuiL replied:
Once again, `quilt` is Loader (or perhaps even `quilt_loader`), `qsl` is QSL; We shouldn't repeat the whole "Use `fabric` as Fabric API's mod ID" mistake

LambdAurora replied:
This is configuration though, very different from a mod id.

My issue with this is if someone who doesn't know what QSL is (which could happen once there's dependency downloading yada yada) it would confuse people.

EnnuiL replied:
It's the same fundamental issue: it tries to simplify too much, and in the result, everyone loses;
There was the whole "We'll call QFAPI "QSL" in order to confuse less people" thing, but in the end, everyone was pretty confused when "QSL" turned out to be QFAPI bundling QSL; We shouldn't lie to the user here, let them know about `qsl`

Patbox replied:
Moved it under `quilt/qsl/registry.toml`. Used `registry` just in case of config being extended in future for non-regsync related options.

EnnuiL reviewed (changes requested):
wait, there's still a card on my sleeves

Patbox committed:
Change config directory

EnnuiL reviewed (approved):
The only thing left is some good ol' testing, other than that, PR approved!

Patbox committed:
Merge branch '1.18' into 1.18

LambdAurora committed:
Merge branch '1.18' into 1.18

LambdAurora committed:
Merge branch '1.18' into 1.18

Patbox committed:
Cache block's loot table (like vanilla/fapi)

Patbox committed:
Missing ;

LambdAurora reviewed (changes requested):
Ok so, little update for everyone!

1. Registry Syncing has been tested with a known desynced modpack, it works-ish.
2. While testing with a Fabric server and a Quilt client we noticed something missing, the module forgot to take care of some id maps, notably the item model one. Causing quite the visual desync: 
![image](https://user-images.githubusercontent.com/12587332/182643886-8ae19210-b4b6-46d7-87ff-ef252279eab9.png)

Thus the module is currently `s: not working`, meanwhile that going into FCP is blocked until this issue is fixed.
Once fixed testing can resume and hopefully get this into FCP ;)

Patbox committed:
Fix some missing ids remaps

Patbox committed:
Apply missing licenses

Patbox committed:
Fix some things not applying to fabric sync

LambdAurora committed:
Cleanup a bit.

Patbox committed:
Patch client only entries sharing ids when not on server

Patbox committed:
Fill rawIdToEntry

LambdAurora committed:
Merge branch '1.18' of github.com:Patbox/quilt-standard-libraries into patbox/1.18

LambdAurora committed:
Add missing contracts.

Leo40Git reviewed (approved):
LGTM!

LambdAurora committed:
Keep order of extra client modded game objects.

LambdAurora reviewed (approved):
Testing has been done with:
 - Quilt servers
 - Fabric servers
 - Quilt clients
 - Fabric clients

and all combination ended up working which means testing is done.

Leo40Git commented:
is it finally time to merge registry sync into qsl
Or, well, have it enter FCP.
